signer/signer_test.go

```diff
@@ -662,3 +663,208 @@ func TestMakeKeyGCPRSA(t *testing.T) {
 		t.Fatalf("MakeKey failed: %v", err)
 	}
 }
+
+var UNLIMITEDBYTESREADERTESTCASES = []struct {
+	requestedBytes                int
+	expectedPkcs11RandReaderBytes []int
+}{
+	{
+		requestedBytes:                1,
+		expectedPkcs11RandReaderBytes: []int{1},
+	},
+	{
+		requestedBytes:                1024,
+		expectedPkcs11RandReaderBytes: []int{1024},
+	},
+	{
+		requestedBytes:                1536,
+		expectedPkcs11RandReaderBytes: []int{1024, 512},
+	},
+	{
+		requestedBytes:                2047,
+		expectedPkcs11RandReaderBytes: []int{1024, 1023},
+	},
+	{
+		requestedBytes:                2048,
+		expectedPkcs11RandReaderBytes: []int{1024, 1024},
+	},
+	{
+		requestedBytes:                2049,
+		expectedPkcs11RandReaderBytes: []int{1024, 1024, 1},
+	},
+	{
+		requestedBytes:                5000,
+		expectedPkcs11RandReaderBytes: []int{1024, 1024, 1024, 1024, 904},
+	},
+}
+
+func TestUnlimitedBytesRandReader(t *testing.T) {
+	ubrr := new(unlimitedBytesRandReader)
+
+	ctrl := gomock.NewController(t)
+	mockCtx := mockpkcs11.NewMockPKCS11Context(ctrl)
+	defer ctrl.Finish()
+
+	slot := uint(0)
+	session := pkcs11.SessionHandle(0)
+	mockCtx.EXPECT().Initialize().Return(nil).Times(1)
+	mockCtx.EXPECT().GetSlotList(true).Return([]uint{slot}, nil).Times(2)
+	mockCtx.EXPECT().GetTokenInfo(slot).Return(pkcs11.TokenInfo{}, nil).Times(1)
+
+	// somewhat surprisingly, a new session is used each time we call
+	// GenerateRandom. It's unclear whether this happens exclusively in tests
+	// or if it also happens in production code
+	expectedSessions := 0
+	for _, testcase := range UNLIMITEDBYTESREADERTESTCASES {
+		expectedSessions += len(testcase.expectedPkcs11RandReaderBytes)
+	}
+	mockCtx.EXPECT().OpenSession(slot, uint(6)).Return(session, nil).Times(expectedSessions)
+
+	mockFactory := mockedPKCS11ContextFactory(mockCtx)
+	crypto11.Configure(&crypto11.PKCS11Config{}, mockFactory)
+	defer crypto11.Close()
+
+	for _, testcase := range UNLIMITEDBYTESREADERTESTCASES {
+		for _, size := range testcase.expectedPkcs11RandReaderBytes {
```

Comment:
I could certainly add this, but unless I'm missing something it would require me to have the mock (or an additional mock) behave differently than the real method, which doesn't seem ideal.

A different thing we could do is to have `unlimitedBytesRandReader.Read` mirror the behaviour of the underlying method, and have `n` always be 0 or `len(data)`. I'm not super excited to do that either, but the more I think about it, the more it actually seems like the most correct thing to do?

Reply:
If it returns 0, it'd return an error, right? You could have the mock return an error.

Just an FYI, there is an [InOrder](https://pkg.go.dev/go.uber.org/mock/gomock#InOrder) api that makes it a lil easier to interleave responses.

Reply:
That's my understanding - yes.

Oh wait, I think I just realized you were asking to make sure that our code behaves correctly -- not that it returns > 0 on error. Yes of course, that's totally doable (and sensible).

Reply:
Oh, this is already covered though? The mock is on `GenerateRandom`, which AFAIK might return a non-zero number of bytes, but also an error. https://github.com/mozilla-services/autograph/pull/1032/files#diff-a43f626c786b027528255923aee643b55513fb5d596a69e082c63dba5b9ccfb9R840-R845 covers this, I believe?

Reply:
SGTM!